src/python/torchdistx/gossip_grad.py

```diff
@@ -0,0 +1,338 @@
+            by default a subgroup is initialized to workers, belonging to the same node.
+            Should be provided together with `num_nodes`.
+            (default: None)
+        num_nodes (int): Number of nodes in a compute environment.
+            Should be provided together with `local_process_group`.
+            By default is initialized to the number of generated local subgroups.
+            (default: None)
+        master_process_group (ProcessGroup): Stores main workers,
+            which are involved in inter-node communication. By default, will be
+            composed from the workers with ranks 0 in the local process group.
+            (default: None)
+
+    """
+
+    def __init__(
+        self,
+        topology=None,
+        local_process_group=None,
+        num_nodes=None,
+        master_process_group=None,
+        proc_per_node=None,
+    ):
+
+        self.topology = topology or Topology.DISSEMINATION
+        if local_process_group is None and num_nodes is None:
+            self.local_process_group, subgroups = dist.new_subgroups()
+            self.num_nodes = len(subgroups)
+        else:
+            if (
+                local_process_group is not None
+                and num_nodes is None
+                or local_process_group is None
+                and num_nodes is not None
+            ):
+                raise ValueError(
+                    "`local_process_group` and `num_nodes` should be provided together."
+                )
+            self.local_process_group = local_process_group
+            if num_nodes < 1:
+                raise ValueError("`num_nodes` should be equal to 1 or more.")
+            self.num_nodes = num_nodes
+
+        if self.num_nodes % 2 != 0 and self.topology == Topology.CUBE:
+            raise ValueError(
+                "Current implementation doesn't support uneven number"
+                " of nodes for CUBE topology."
+            )
+
+        super().__init__(self.local_process_group)
+        self.proc_per_node = (
+            proc_per_node
+            if proc_per_node is not None
+            else self.local_process_group.size()
+        )
+        if self.proc_per_node < 1:
+            raise ValueError("`proc_per_node` should be equal to 1 or more.")
+
+        self.master_process_group = (
+            master_process_group
+            if master_process_group is not None
+            else self._create_master_group()
+        )
+
+        self.topologies = self._generate_topologies()
+        self.cur_topology = next(self.topologies)
+
+        # For `num_nodes` != power of 2 `gossip_period` should still be an int.
+        # If we only have 1 node, `gossip_period` should be equal to 1.
+        self.gossip_period = max(1, math.ceil(math.log(self.num_nodes, 2)))
+        self.iter = 0
+
+        # Set device for `dist.batch_isend_irecv`
+        self.rank = torch.cuda.current_device()
+        torch.cuda.set_device(self.rank)
```

Comment:
+1 to Rohan's comment above. 

Besides, `set_device` will change global default device. Is this side effect OK?

Reply:
@rohan-varma and @mrshenli I see, this part needs to be changed. So I will remove `set_device` and add  requirements for `batch_isend_irecv` into the doc. Current requirements are : 

> Note that when this API is used with the NCCL PG backend, users must set the current GPU device with `torch.cuda.set_device`, otherwise it will lead to unexpected hang issues.